File: ios/Sources/KlinikAPI/ProtocolsAPI.swift
import Foundation
import KlinikCore

/// A document the assistant is allowed to quote (spec M4).
public struct ProtocolDocument: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let title: String
    /// Restricts retrieval to patients who had this procedure. Nil means every
    /// patient, which is right for a general FAQ and wrong for a discharge
    /// instruction that belongs to one operation.
    public let procedureType: String?
    public let language: String
    public let content: String
    public let version: Int
    public let isActive: Bool
    public let createdAt: Date
}

public struct ProtocolSummary: Decodable, Sendable, Equatable, Identifiable {
    public let document: ProtocolDocument
    /// How many passages it was split into. Zero means the assistant has
    /// nothing to retrieve from it.
    public let chunks: Int
    /// False when no embedding provider was configured at upload time — the
    /// document is stored and the assistant cannot find it.
    public let embedded: Bool

    public var id: String { document.id }

    /// Whether this document can actually answer anything today.
    public var isUsable: Bool { document.isActive && embedded && chunks > 0 }
}

public struct ProtocolsAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func list(includeInactive: Bool = false) async throws -> [ProtocolSummary] {
        try await client.send(
            Endpoint(
                method: .get,
                path: "protocols",
                query: includeInactive ? ["includeInactive": "true"] : [:]
            ),
            as: [ProtocolSummary].self
        )
    }

    public func upload(
        title: String,
        content: String,
        procedureType: String?,
        language: String
    ) async throws -> ProtocolSummary {
        try await client.send(
            Endpoint(
                method: .post,
                path: "protocols",
                body: try JSONEncoder.klinik.encode(
                    UploadBody(
                        title: title,
                        content: content,
                        procedureType: procedureType,
                        language: language
                    )
                )
            ),
            as: ProtocolSummary.self
        )
    }

    /// Retires a document. The assistant stops quoting it; the record stays,
    /// because an answer it gave last week was given from something.
    public func retire(_ documentId: String) async throws -> ProtocolDocument {
        try await client.send(
            Endpoint(method: .delete, path: "protocols/\(documentId)"),
            as: ProtocolDocument.self
        )
    }

    private struct UploadBody: Encodable {
        let title: String
        let content: String
        let procedureType: String?
        let language: String
    }
}

File: android/core/network/src/main/kotlin/xyz/klinik/network/ProtocolsApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The documents the assistant is allowed to answer from (spec M4).
 *
 * This is the only place the clinic decides what the FAQ assistant knows. A
 * document uploaded here is chunked and embedded on the way in, and the
 * assistant answers from these and nothing else — which is why `embedded`
 * being false matters enough to be its own field: the document is stored and
 * the assistant cannot see it.
 */
@Serializable
data class ProtocolDocument(
    val id: String,
    val title: String,
    /** Restricts retrieval to patients who had this procedure. */
    val procedureType: String? = null,
    val language: String,
    val isActive: Boolean = false,
    val createdAt: String,
)

@Serializable
data class ProtocolSummary(
    val document: ProtocolDocument,
    /** How many pieces the document was split into. */
    val chunks: Int = 0,
    /**
     * False when no embedding provider was configured.
     *
     * The document is stored either way, and the assistant cannot retrieve it
     * — so a list that did not show this would show a clinic a protocol its
     * assistant has never read.
     */
    val embedded: Boolean = false,
) {
    val isUsable: Boolean get() = document.isActive && embedded && chunks > 0
}

@Serializable
private data class UploadProtocolBody(
    val title: String,
    val content: String,
    val procedureType: String? = null,
    val language: String = "tr",
)

class ProtocolsApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    suspend fun all(): List<ProtocolSummary> =
        decode(client.send(Endpoint(HttpMethod.GET, "protocols")))

    /** Uploads one. The text is chunked and embedded on the way in. */
    suspend fun upload(
        title: String,
        content: String,
        procedureType: String? = null,
        language: String = "tr",
    ): ProtocolSummary =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.POST,
                    "protocols",
                    body = json.encodeToString(
                        UploadProtocolBody.serializer(),
                        UploadProtocolBody(title, content, procedureType, language),
                    ),
                ),
            ),
        )

    /** Withdraws one, so the assistant stops answering from it. */
    suspend fun remove(documentId: String) {
        client.send(Endpoint(HttpMethod.DELETE, "protocols/$documentId"))
    }

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}
